### PR TITLE
DM-42189: Serve multiple Butler repositories from a single server

### PR DESCRIPTION
**Pull request by dhirving**
In most environments, users will access multiple Butler repositories (multiple data releases, prompt processing, etc.).  To avoid the need to instantiate a separate service for every repository, allow multiple repositories to be hosted from a single server.

This uses the same DAF_BUTLER_REPOSITORY_INDEX configuration system as DirectButler.

## Checklist

- [x] ran Jenkins
- [ ] added a release note for user-visible changes to `doc/changes`

### CODE REVIEW

File: python/lsst/daf/butler/remote_butler/server/_factory.py [hunk 1]
```diff
@@ -25,22 +25,31 @@
...
+    """Class for instantiating per-request dependencies, following the pattern
+    in `SQR-072 <https://sqr-072.lsst.io/>`_.
 
     Parameters
     ----------
-    butler : `DirectButler`
-        Butler to use.
+    repository : `str`
+        The label of the Butler repository requested by the user.
+
```
Comment (timj) [inline python/lsst/daf/butler/remote_butler/server/_factory.py]: Numpydoc doesn't need the blank line here.

File: tests_integration/test_docker_container.py [hunk 2]
```diff
@@ -24,21 +24,28 @@ def _run_server_docker():
 
         port = 8080
         butler_root = "/butler_root"
+
+        # Set up a repository index file to be read in by the server
+        index_filename = "repo_index.yaml"
+        repo_name = "testserver"
+        with open(os.path.join(temp_dir, index_filename), "wb") as fh:
+            fh.write(f"{repo_name}: {butler_root}\n".encode())
```
Comment (timj) [inline tests_integration/test_docker_container.py:32]:
```suggestion
            yaml.dump({repo_name: butler_root}, stream=fh)
```
?

**Review by timj — APPROVED** [review]
Looks good. Is the `/repo` in the path really necessary?

**Comment by dhirving** [comment]
> Looks good. Is the `/repo` in the path really necessary?

Yeah, it leaves a free namespace where we can put per-server endpoints, like "list_available_repositories" or "health_check".  Otherwise we'd have to worry about repository labels colliding with endpoint names.  (Not sure but I wouldn't be surprised if FastAPI doesn't even support having a `/{placeholder}` that potentially collides with another name.)

**Pull request merged**